music/demos/callback_player.ts

```diff
@@ -0,0 +1,213 @@
+      [document.getElementById('d_sharp')],
+      [document.getElementById('top_e'),
+       document.getElementById('left_bottom_e'),
+       document.getElementById('bottom_e')],
+      [document.getElementById('top_f'),
+       document.getElementById('bottom_f'),
+       document.getElementById('right_bottom_f')],
+      [document.getElementById('f_sharp')],
+      [document.getElementById('top_g'),
+       document.getElementById('left_bottom_g'),
+       document.getElementById('bottom_g'),
+       document.getElementById('right_bottom_g')],
+      [document.getElementById('g_sharp')],
+      [document.getElementById('top_a'),
+       document.getElementById('left_bottom_a'),
+       document.getElementById('bottom_a'),
+       document.getElementById('right_bottom_a')],
+      [document.getElementById('a_sharp')],
+      [document.getElementById('top_b'),
+       document.getElementById('left_bottom_b'),
+       document.getElementById('bottom_b')]
+    ];
+    this.drumPitchToClass = new Map<number, number>();
+    for (let c = 0; c < mm.DEFAULT_DRUM_PITCH_CLASSES.length; ++c) {
+      mm.DEFAULT_DRUM_PITCH_CLASSES[c].forEach((p) => {
+        this.drumPitchToClass.set(p, c);
+      });
+    }
+    this.beatPos = -1;
+    this.currentDrum = -1;
+    this.greyOut(this.clickDivs, -1);
+    this.greyOut(this.drumDivs, -1);
+  }
+
+  greyOut(divs: any, ignorePos: number) {
+    for (let i = 0; i < divs.length; ++i) {
+      if (i == ignorePos) continue;
+      divs[i].style.background = 'grey';
+    }
+  }
+
+  colorDrums(n: mm.NoteSequence.INote) {
+    if (n.pitch == mm.HI_CLICK_PITCH) {
+      this.beatPos = 0;
+      this.clickDivs[this.beatPos].style.background = 'red';
+      setTimeout(() => {
+        this.clickDivs[this.beatPos].style.background = 'grey';
+      }, 300);
+    } else if (n.pitch == mm.LO_CLICK_PITCH) {
+      if (this.beatPos < this.clickDivs.length - 1) {
+        ++this.beatPos;
+      }
+      this.clickDivs[this.beatPos].style.background = 'blue';
+      setTimeout(() => {
+        this.clickDivs[this.beatPos].style.background = 'grey';
+      }, 300);
+    } else {
+      this.currentDrum = this.drumPitchToClass.get(n.pitch);
+      this.drumDivs[this.currentDrum].style.background = 'green';
+      setTimeout(() => {
+        this.drumDivs[this.currentDrum].style.background = 'grey';
+      }, 100);
+    }
+  }
+
+  colorKeys(n: mm.NoteSequence.INote) {
+    let relativePitch = (n.pitch - 60) % 12;
+    for (let i = 0; i < this.keyDivs[relativePitch].length; ++i) {
+      this.keyDivs[relativePitch][i].style.background = 'red';
+      let color = 'white';
+      if (relativePitch == 1 || relativePitch == 3 || relativePitch == 6 ||
+          relativePitch == 8 || relativePitch == 10) {
+        color = 'black';
+      }
+      setTimeout(() => {
+        this.keyDivs[relativePitch][i].style.background = color;
+      }, 100);
+    }
+  }
+
+  run(t: number, n: mm.NoteSequence.INote) {
+    if (n.isDrum) {
```

Comment:
To get the visualization to sync up exactly with the sound, you can schedule it with [Tone.Draw](https://tonejs.github.io/docs/r12/Draw). Otherwise it'll show early, as Tone events are fired slightly before the scheduled time.

Reply:
thanks! done